### ISSUE TITLE
Large Files uploading to dropbox restarting from 0

### ISSUE THREAD

**Issue by zenjabba** [ncw/rclone#1806]
```
file1.mkv: 17% done, 7.613 MBytes/s, ETA: 14m3s
file1.mkv:  0% done, 17.252 MBytes/s, ETA: 7m30s
file1.mkv:  1% done, 946.785 kBytes/s, ETA: 2h18m49s
file1.mkv: 17% done, 2.200 MBytes/s, ETA: 48m38s
file1.mkv:  2% done, 2.071 MBytes/s, ETA: 1h1m34s
```

It's like dropbox keeps being told to "start again". This can get very tiring when you cannot upload large files (40GB)

**Comment by zenjabba** [comment]
found this error `Post https://content.dropboxapi.com/2/files/upload_session/append_v2: http: server closed idle connection`

**Comment by zenjabba** [comment]
Found more information

2017/11/08 03:02:05 DEBUG : pacer: low level retry 1/1 (error Post https://content.dropboxapi.com/2/files/upload_session/append_v2: unexpected EOF)
2017/11/08 03:02:05 DEBUG : Jellyfish Eyes (2013)/Jellyfish.Eyes.2013.Criterion.Collection.BluRay.Remux.1080p.AVC.DTS-NCmt.mkv: Received error: upload failed: Post https://content.dropboxapi.com/2/files/upload_session/append_v2: unexpected EOF - low level retry 1/10
2017/11/08 03:02:05 DEBUG : pacer: Reducing sleep to 15ms

Basically on the first error, it restarts from the start, rather than uploading the same "block"

**Comment by zenjabba** [comment]
I'm assuming `// FIXME buffer chunks for retries in upload` means that this is a known problem... ARGH!

**Comment by ncw** [comment]
Yes you are right, that is the all important FIXME!

At the moment we send 128MB chunks but we don't buffer them, so if there is any sort of error then we have to start again.

I've had a go at adding buffering - can you have a go and let me know if it works!  I know it works in the non error case, but if you can see it working retrying a chunk that would be useful confirmation.

https://pub.rclone.org/v1.38-094-gb11f2dd2-dropbox-buffer%CE%B2/

It might be that the 128M default buffer size is too big now...

**Comment by zenjabba** [comment]
This one is broken too..


2017/11/08 13:17:54 INFO  : Dropbox root 'mjf14/backups/source/film': Modify window is 1s
2017/11/08 13:18:52 ERROR : Dead Man Down (2013)/Dead.Man.Down.2013.1080p.BluRay.REMUX.AVC.DTS-HD.MA.5.1-EPSiLON.mkv: Failed to copy: upload failed: unexpected end of JSON input
2017/11/08 13:18:54 INFO  :
Transferred:   110 MBytes (1.806 MBytes/s)
Errors:                 1
Checks:                 0
Transferred:            0
Elapsed time:      1m0.9s
Transferring:
 * Atomic Blonde (2017)/Atomic.Blonde.2017.BluRay.1080p.DTS-X.7.1.AVC.REMUX-FraMeSToR.mkv

when you get a failure, it just skips it and moves to the next one.

**Comment by zenjabba** [comment]
I needed to reduce the buffer to 10M which worked better also.

**Comment by zenjabba** [comment]
I can confirm this has NOT fixed the issue.

**Comment by zenjabba** [comment]
@ncw I can give you access to the system to test if that will help.

**Comment by zenjabba** [comment]
Update. This version is also presenting this error

```2017/11/09 21:36:46 ERROR : file.mkv Failed to copy: upload failed: lookup_failed/incorrect_offset/.```

**Comment by ncw** [comment]
> when you get a failure now, it just skips it and moves to the next one. This isn't the expected outcome :)

I really need to see a complete log with `-vv` of the failure.  Feel free to search and replace the file names if you want.

**Comment by zenjabba** [comment]
```2017/11/10 13:51:33 DEBUG : Using config file from "/root/.config/rclone/rclone.conf"
2017/11/10 13:51:33 DEBUG : rclone: Version "v1.38-094-gb11f2dd2-dropbox-bufferβ" starting with parameters ["rclone" "copy" "" "" "--checkers=6" "--transfers=10" "--dropbox-chunk-size=15M" "-vv" "--log-file=/tmp/rclone.copyfile"]
2017/11/10 13:51:52 DEBUG : ABM.mkv: Uploading chunk 1/770
2017/11/10 13:51:59 DEBUG : ABM.mkv: Uploading chunk 2/770
2017/11/10 13:52:02 DEBUG : ABM.mkv: Uploading chunk 3/770
2017/11/10 13:52:06 DEBUG : ABM.mkv: Uploading chunk 4/770
2017/11/10 13:52:10 DEBUG : ABM.mkv: Uploading chunk 5/770
2017/11/10 13:52:13 DEBUG : ABM.mkv: Uploading chunk 6/770
2017/11/10 13:52:17 DEBUG : ABM.mkv: Uploading chunk 7/770
2017/11/10 13:52:21 DEBUG : ABM.mkv: Uploading chunk 8/770
2017/11/10 13:52:28 DEBUG : ABM.mkv: Uploading chunk 10/770
2017/11/10 13:52:32 DEBUG : ABM.mkv: Uploading chunk 11/770
2017/11/10 13:52:36 DEBUG : ABM.mkv: Uploading chunk 12/770
2017/11/10 13:52:39 DEBUG : ABM.mkv: Uploading chunk 13/770
2017/11/10 13:52:42 DEBUG : ABM.mkv: Uploading chunk 14/770
2017/11/10 13:52:46 DEBUG : ABM.mkv: Uploading chunk 15/770
2017/11/10 13:52:50 DEBUG : ABM.mkv: Uploading chunk 16/770
2017/11/10 13:52:53 DEBUG : ABM.mkv: Uploading chunk 17/770
2017/11/10 13:52:56 DEBUG : ABM.mkv: Uploading chunk 18/770
2017/11/10 13:53:00 DEBUG : ABM.mkv: Uploading chunk 19/770
2017/11/10 13:53:04 DEBUG : ABM.mkv: Uploading chunk 20/770
2017/11/10 13:53:07 DEBUG : ABM.mkv: Uploading chunk 21/770
2017/11/10 13:53:10 DEBUG : ABM.mkv: Uploading chunk 22/770
2017/11/10 13:53:14 DEBUG : ABM.mkv: Uploading chunk 23/770
2017/11/10 13:53:17 DEBUG : ABM.mkv: Uploading chunk 24/770
2017/11/10 13:53:22 DEBUG : ABM.mkv: Uploading chunk 25/770
2017/11/10 13:53:26 DEBUG : ABM.mkv: Uploading chunk 26/770
2017/11/10 13:53:30 DEBUG : ABM.mkv: Uploading chunk 27/770
2017/11/10 13:53:34 DEBUG : ABM.mkv: Uploading chunk 28/770
2017/11/10 13:53:39 DEBUG : ABM.mkv: Uploading chunk 29/770
2017/11/10 13:53:43 DEBUG : ABM.mkv: Uploading chunk 30/770
2017/11/10 13:53:48 DEBUG : ABM.mkv: Uploading chunk 31/770
2017/11/10 13:53:52 DEBUG : ABM.mkv: Uploading chunk 32/770
2017/11/10 13:53:56 DEBUG : ABM.mkv: Uploading chunk 33/770
2017/11/10 13:53:59 DEBUG : ABM.mkv: Uploading chunk 34/770
2017/11/10 13:54:03 DEBUG : ABM.mkv: Uploading chunk 35/770
2017/11/10 13:54:07 DEBUG : ABM.mkv: Uploading chunk 36/770
2017/11/10 13:54:10 DEBUG : ABM.mkv: Uploading chunk 37/770
2017/11/10 13:54:10 DEBUG : pacer: Rate limited, increasing sleep to 20ms
2017/11/10 13:54:10 DEBUG : pacer: low level retry 1/10 (error Post https://content.dropboxapi.com/2/files/upload_session/append_v2: unexpected EOF)
2017/11/10 13:54:10 DEBUG : pacer: Rate limited, increasing sleep to 40ms
2017/11/10 13:54:10 DEBUG : pacer: low level retry 2/10 (error Post https://content.dropboxapi.com/2/files/upload_session/append_v2: unexpected EOF)
2017/11/10 13:54:11 DEBUG : pacer: Reducing sleep to 30ms
2017/11/10 13:54:11 DEBUG : pacer: Rate limited, increasing sleep to 60ms
2017/11/10 13:54:11 DEBUG : pacer: low level retry 3/10 (error Post https://content.dropboxapi.com/2/files/upload_session/append_v2: unexpected EOF)
2017/11/10 13:54:11 DEBUG : pacer: Reducing sleep to 45ms
2017/11/10 13:54:11 DEBUG : pacer: Rate limited, increasing sleep to 90ms
2017/11/10 13:54:11 DEBUG : pacer: low level retry 4/10 (error Post https://content.dropboxapi.com/2/files/upload_session/append_v2: unexpected EOF)
2017/11/10 13:54:11 DEBUG : pacer: Rate limited, increasing sleep to 180ms
2017/11/10 13:54:11 DEBUG : pacer: low level retry 5/10 (error Post https://content.dropboxapi.com/2/files/upload_session/append_v2: unexpected EOF)
2017/11/10 13:54:11 DEBUG : pacer: Rate limited, increasing sleep to 360ms
2017/11/10 13:54:11 DEBUG : pacer: low level retry 6/10 (error Post https://content.dropboxapi.com/2/files/upload_session/append_v2: unexpected EOF)
2017/11/10 13:54:11 DEBUG : pacer: Rate limited, increasing sleep to 720ms
2017/11/10 13:54:11 DEBUG : pacer: low level retry 7/10 (error Post https://content.dropboxapi.com/2/files/upload_session/append_v2: unexpected EOF)
2017/11/10 13:54:11 DEBUG : pacer: Reducing sleep to 540ms
2017/11/10 13:54:12 DEBUG : pacer: Rate limited, increasing sleep to 1.08s
2017/11/10 13:54:12 DEBUG : pacer: low level retry 8/10 (error Post https://content.dropboxapi.com/2/files/upload_session/append_v2: unexpected EOF)
2017/11/10 13:54:12 DEBUG : pacer: Reducing sleep to 810ms
2017/11/10 13:54:12 DEBUG : pacer: Reducing sleep to 607.5ms
2017/11/10 13:54:13 DEBUG : pacer: Rate limited, increasing sleep to 1.215s
2017/11/10 13:54:13 DEBUG : pacer: low level retry 9/10 (error Post https://content.dropboxapi.com/2/files/upload_session/append_v2: unexpected EOF)
2017/11/10 13:54:13 DEBUG : pacer: Reducing sleep to 911.25ms
2017/11/10 13:54:13 DEBUG : pacer: Reducing sleep to 683.4375ms
2017/11/10 13:54:14 DEBUG : pacer: Reducing sleep to 512.578125ms
2017/11/10 13:54:14 DEBUG : pacer: Reducing sleep to 384.433593ms
2017/11/10 13:54:15 DEBUG : pacer: Reducing sleep to 288.325194ms
2017/11/10 13:54:15 DEBUG : pacer: Rate limited, increasing sleep to 576.650388ms
2017/11/10 13:54:15 DEBUG : pacer: low level retry 10/10 (error Post https://content.dropboxapi.com/2/files/upload_session/append_v2: unexpected EOF)
2017/11/10 13:54:15 DEBUG : ABM.mkv: Received error: upload failed: Post https://content.dropboxapi.com/2/files/upload_session/append_v2: unexpected EOF - low level retry 1/10
2017/11/10 13:54:15 DEBUG : pacer: Reducing sleep to 432.487791ms
2017/11/10 13:54:16 DEBUG : ABM.mkv: Uploading chunk 1/770
2017/11/10 13:54:23 DEBUG : ABM.mkv: Uploading chunk 2/770
2017/11/10 13:54:28 DEBUG : ABM.mkv: Uploading chunk 3/770
2017/11/10 13:54:32 DEBUG : ABM.mkv: Uploading chunk 4/770
2017/11/10 13:54:37 DEBUG : ABM.mkv: Uploading chunk 5/770
2017/11/10 13:54:41 DEBUG : ABM.mkv: Uploading chunk 6/770
2017/11/10 13:54:47 DEBUG : ABM.mkv: Uploading chunk 7/770

**Comment by ncw** [comment]
I've figured out what is causing the "unexpected end of JSON input" errors - it is a bug in the dropbox SDK: https://github.com/dropbox/dropbox-sdk-go-unofficial/issues/36

I've thought of a good work-around which I'm going to try - namely retrying all the errors after the first chunk has been uploaded.

**Comment by ncw** [comment]
I have spent some time investigating this and I've made the large file uploader much more robust.  It can retry segments and it will try harder not to give up.  I've tested it uploading a couple of TB of files with no errors now.

https://beta.rclone.org/v1.38-178-g8357a82e/ (uploaded in 15-30 mins)
